### PR TITLE
Remove deprecated arguments with regard to `LightGBM>=4.0`

### PR DESCRIPTION
**Pull request by Alnusjaponica**
## Motivation
Make behavior of `_LightGBMBaseTuner`, `LightGBMTuner` and `LightGBMTunerCV` consistent with the lates `LightGBM`.

## Description of the changes
Remove deprecated arguments in `LightgGBM>=4.0` from `_LightGBMBaseTuner`, `LightGBMTuner` and `LightGBMTunerCV`

### CODE REVIEW

**Comment by toshihikoyanase** [comment]
@not522 Could you review this PR, please?

**Comment by toshihikoyanase** [comment]
> Remove deprecated arguments in `LightgGBM>=4.0` from `_LightGBMBaseTuner`, `LightGBMTuner` and `LightGBMTunerCV`.

Sounds relevant since the integration modules basically align with the latest version of the target libraries.

**Review by toshihikoyanase — COMMENTED** [review]
I tried to follow the [LightGBM reference](https://lightgbm.readthedocs.io/en/v4.0.0/Python-API.html) and the [release notes](https://github.com/microsoft/LightGBM/releases/tag/v4.0.0), and I found that it difficult to identify which arguments were deprecated in v4.0.0.

Could you show me the references the change in this PR, please?
- The default value of `callbacks`
- Deprecation of `fobj`
- Deprecation of `early_stopping_rounds`
- Deprecation of `verbose_eval`
- Deprecation of `evals_result`
- Deprecation of `learning_rates`

**Comment by Alnusjaponica** [comment]
Although there are no clear issue corresponding to the deprecation, they were remove by the following PRs.
I still haven't fully grasped the concept of "the default value of callbacks." I would greatly appreciate it if you could kindly provide further clarification.
`fobj`: https://github.com/microsoft/LightGBM/pull/5052
`early_stopping_rounds`: https://github.com/microsoft/LightGBM/pull/4908
`verbose_eval`: https://github.com/microsoft/LightGBM/pull/4878
`evals_result`: https://github.com/microsoft/LightGBM/pull/4882
`learning_rates`: https://github.com/microsoft/LightGBM/pull/4831

**Review by not522 — APPROVED** [review]
Thank you. LGTM.

File: optuna/integration/_lightgbm_tuner/optimize.py [hunk 1]
```diff
@@ -358,14 +358,11 @@ def __init__(
         self,
         params: dict[str, Any],
         train_set: "lgb.Dataset",
-        callbacks: list[Callable[..., Any]],
+        callbacks: list[Callable[..., Any]] | None = None,
         num_boost_round: int = 1000,
-        fobj: Callable[..., Any] | None = None,
```
Comment (toshihikoyanase) [inline optuna/integration/_lightgbm_tuner/optimize.py:363]: [Notes] See https://github.com/microsoft/LightGBM/pull/5052. Use `params['objective']` instead.

File: optuna/integration/_lightgbm_tuner/optimize.py [hunk 2]
```diff
@@ -358,14 +358,11 @@ def __init__(
...
         feval: Callable[..., Any] | None = None,
         feature_name: str = "auto",
         categorical_feature: str = "auto",
-        early_stopping_rounds: int | None = None,
```
Comment (toshihikoyanase) [inline optuna/integration/_lightgbm_tuner/optimize.py:367]: [Notes] See https://github.com/microsoft/LightGBM/pull/4908. Use `callbacks=[lgb.early_stopping(stopping_rounds=N)]` instread.

File: optuna/integration/_lightgbm_tuner/optimize.py [hunk 3]
```diff
@@ -358,14 +358,11 @@ def __init__(
...
-        verbose_eval: bool | int | str | None = None,
```
Comment (toshihikoyanase) [inline optuna/integration/_lightgbm_tuner/optimize.py:368]: See https://github.com/microsoft/LightGBM/pull/4878. Use `callbacks=[lgb.log_evaluation(10)]` instead.

File: optuna/integration/_lightgbm_tuner/optimize.py [hunk 4]
```diff
@@ -778,14 +758,9 @@ def __init__(
         num_boost_round: int = 1000,
         valid_sets: "VALID_SET_TYPE" | None = None,
         valid_names: Any | None = None,
-        fobj: Callable[..., Any] | None = None,
         feval: Callable[..., Any] | None = None,
         feature_name: str = "auto",
         categorical_feature: str = "auto",
-        early_stopping_rounds: int | None = None,
-        evals_result: dict[Any, Any] | None = None,
```
Comment (toshihikoyanase) [inline optuna/integration/_lightgbm_tuner/optimize.py:786]: [Notes] See https://github.com/microsoft/LightGBM/pull/4882. Use `callbacks=[.record_evaluation(evals_result)]` instead.

File: optuna/integration/_lightgbm_tuner/optimize.py [hunk 5]
```diff
@@ -778,14 +758,9 @@ def __init__(
...
-        verbose_eval: bool | int | str | None = "warn",
-        learning_rates: list[float] | None = None,
```
Comment (toshihikoyanase) [inline optuna/integration/_lightgbm_tuner/optimize.py:788]: [Notes] See https://github.com/microsoft/LightGBM/pull/4831. Use `callbacks=[lgb.reset_parameter(learning_rate=...)]` instead.

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your investigation.
The change looks solid to me.

**Comment by toshihikoyanase** [comment]
Please note that I believe adding a version constraint for `lightgbm` may not be necessary, as it keeps backward compatibility for the majority of users. This is particularly true for those utilizing the LightGBM tuner with the default options.

**Pull request merged**